File: src/main/java/me/partlysanestudios/partlysaneskies/render/gui/components/PSSToggle.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.render.gui.components;

import gg.essential.elementa.UIComponent;
import gg.essential.elementa.components.UIBlock;
import gg.essential.elementa.components.UIImage;
import gg.essential.elementa.constraints.CenterConstraint;
import gg.essential.elementa.constraints.PixelConstraint;
import gg.essential.elementa.constraints.XConstraint;
import gg.essential.elementa.constraints.YConstraint;
import gg.essential.elementa.events.UIClickEvent;
import gg.essential.elementa.utils.ObservableList;
import me.partlysanestudios.partlysaneskies.features.themes.ThemeManager;
import net.minecraft.util.ResourceLocation;

import java.awt.*;
import java.util.function.Consumer;

public class PSSToggle {
    static final ResourceLocation UNSELECTED_IMAGE_PATH = new ResourceLocation("partlysaneskies" , "textures/gui/unselected_toggle.png");
    static final ResourceLocation SELECTED_IMAGE_PATH = new ResourceLocation("partlysaneskies" , "textures/gui/selected_toggle.png");

    float width;
    float height;

    XConstraint xConstraint;
    YConstraint yConstraint;

    boolean isSelected;

    UIBlock backgroundBlock;
    UIImage buttonTexture;

    public PSSToggle() {
        this.isSelected = false;
        this.backgroundBlock = (UIBlock) new UIBlock()
                .setColor(new Color(0, 0,0,0));

        this.buttonTexture = (UIImage) ThemeManager.getCurrentToggleUIImage(false)
                .setChildOf(backgroundBlock);

        this.buttonTexture.onMouseClickConsumer(s -> {
        });
    }

    public PSSToggle toggleState() {
        return this.setState(!this.isSelected);

    }

    public PSSToggle setState(boolean state) {
        this.isSelected = state;

        return this.updateState();
    }

    public PSSToggle updateState() {

        ObservableList<UIComponent> children = buttonTexture.getChildren();

        backgroundBlock.removeChild(buttonTexture);

        if (this.isSelected) {
            this.buttonTexture = (UIImage) ThemeManager.getCurrentToggleUIImage(true)
                    .setWidth(new PixelConstraint(this.width))
                    .setHeight(new PixelConstraint(this.height))
                    .setX(new CenterConstraint())
                    .setY(new CenterConstraint())
                    .setChildOf(this.backgroundBlock);
        }
        else {
            this.buttonTexture = (UIImage) ThemeManager.getCurrentToggleUIImage(false)
                    .setWidth(new PixelConstraint(this.width))
                    .setHeight(new PixelConstraint(this.height))
                    .setX(new CenterConstraint())
                    .setY(new CenterConstraint())
                    .setChildOf(this.backgroundBlock);
        }

        for (UIComponent child : children) {
            child.setChildOf(this.buttonTexture);
        }
        return this;
    }

    public PSSToggle setHeight(float height) {
        backgroundBlock.setHeight(new PixelConstraint(height));
        buttonTexture.setHeight(new PixelConstraint(height))
                .setX(new CenterConstraint())
                .setY(new CenterConstraint());

        this.height = height;

        return this;
    }

    public PSSToggle setWidth(float width) {
        backgroundBlock.setWidth(new PixelConstraint(width));
        buttonTexture.setWidth(new PixelConstraint(width))
                .setX(new CenterConstraint())
                .setY(new CenterConstraint());

        this.width = width;

        return this;
    }

    public PSSToggle setX(XConstraint xPos) {
        backgroundBlock.setX(xPos);
        buttonTexture.setX(new CenterConstraint());

        this.xConstraint = xPos;

        return this;
    }

    public PSSToggle setY(YConstraint yPos) {
        backgroundBlock.setY(yPos);
        buttonTexture.setY(new CenterConstraint());

        this.yConstraint = yPos;

        return this;
    }

//    This method fires when the button has been clicked
    public PSSToggle onMouseClickConsumer(Consumer<UIClickEvent> method) {
        backgroundBlock.onMouseClickConsumer(method);

        return this;
    }

    public UIComponent getComponent() {
        return buttonTexture;
    }

    public PSSToggle insertComponentBeforeBackground(UIComponent component) {
        backgroundBlock.insertChildBefore(component, buttonTexture);

        return this;
    }

    public PSSToggle setChildOf(UIComponent parent) {
        backgroundBlock.setChildOf(parent);

        return this;
    }

    public boolean getState() {
        return isSelected;
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/render/gui/components/PSSToggle.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//
package me.partlysanestudios.partlysaneskies.render.gui.components

import gg.essential.elementa.UIComponent
import gg.essential.elementa.components.UIBlock
import gg.essential.elementa.components.UIImage
import gg.essential.elementa.constraints.CenterConstraint
import gg.essential.elementa.constraints.HeightConstraint
import gg.essential.elementa.constraints.WidthConstraint
import gg.essential.elementa.constraints.XConstraint
import gg.essential.elementa.constraints.YConstraint
import gg.essential.elementa.dsl.childOf
import gg.essential.elementa.dsl.constrain
import gg.essential.elementa.dsl.percent
import gg.essential.elementa.events.UIClickEvent
import me.partlysanestudios.partlysaneskies.features.themes.ThemeManager.getCurrentToggleUIImage
import java.awt.Color
import java.util.function.Consumer

class PSSToggle {
    private var state = false
    private var backgroundBlock: UIBlock =
        UIBlock()
            .setColor(Color(0, 0, 0, 0)) as UIBlock

    private var buttonTexture: UIImage =
        getCurrentToggleUIImage(false).constrain {
            x = CenterConstraint()
            y = CenterConstraint()
            width = 100.percent
            height = 100.percent
        } childOf backgroundBlock

    fun toggleState(): PSSToggle = setState(!state)

    fun setState(state: Boolean): PSSToggle {
        this.state = state
        return updateState()
    }

    fun getState(): Boolean = this.state

    private fun updateState(): PSSToggle {
        val children = buttonTexture.children
        backgroundBlock.removeChild(buttonTexture)

        buttonTexture =
            if (state) {
                getCurrentToggleUIImage(true)
            } else {
                getCurrentToggleUIImage(false)
            }.constrain {
                width = 100.percent
                height = 100.percent
                x = CenterConstraint()
                y = CenterConstraint()
            }

        backgroundBlock.insertChildAt(buttonTexture, 0)

        for (child in children) {
            child.setChildOf(buttonTexture)
        }
        return this
    }

    fun setHeight(height: HeightConstraint): PSSToggle {
        backgroundBlock.setHeight(height)
        return this
    }

    fun setWidth(width: WidthConstraint): PSSToggle {
        backgroundBlock.setWidth(width)
        return this
    }

    fun setX(xPos: XConstraint): PSSToggle {
        backgroundBlock.setX(xPos)
        return this
    }

    fun setY(yPos: YConstraint): PSSToggle {
        backgroundBlock.setY(yPos)
        return this
    }

    // This method fires when the button has been clicked
    fun onMouseClickConsumer(method: Consumer<UIClickEvent>): PSSToggle {
        backgroundBlock.onMouseClickConsumer(method)
        return this
    }

    val component: UIComponent
        get() = backgroundBlock

    fun insertComponentBeforeBackground(component: UIComponent): PSSToggle {
        backgroundBlock.insertChildBefore(component, buttonTexture)
        return this
    }

    fun hide(instantly: Boolean = false) {
        backgroundBlock.hide(instantly)
        buttonTexture.hide(instantly)
    }

    fun unhide(instantly: Boolean = false) {
        backgroundBlock.unhide(instantly)
        buttonTexture.unhide(instantly)
    }

    fun setChildOf(parent: UIComponent): PSSToggle {
        backgroundBlock.setChildOf(parent)
        return this
    }
}
